walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -36,16 +36,34 @@ public int compare(Event a, Event b) {
       };
 
   /**
-   * Returns a list of time periods in which the meeting, specified by request, could happen.
+   * Returns a list of time periods in which the meeting, specified by request, could happen. If one
+   * or more time slots exists so that both mandatory and optional attendees can attend, it returns
+   * those time slots. Otherwise, it returns the time slots that fit just the mandatory attendees.
    *
    * @param eventsCollection the events we know about
    * @param request information about the meeting, including attendees, optional attendees, and how
    *     long it needs to be
    */
   public Collection<TimeRange> query(Collection<Event> eventsCollection, MeetingRequest request) {
-    ArrayList<Event> events =
-        getRelevantEvents(
-            new HashSet<String>(request.getAttendees()), new ArrayList<Event>(eventsCollection));
+    Collection<TimeRange> withOptionalAttendees =
+        getMeetingTimes(eventsCollection, request, /*includeOptionalAttendees=*/ true);
+
+    // Special case: if no mandatory attendees and optional attendees' schedules cannot fit in a
+    // meeting, no meeting times are possible.
+    return !withOptionalAttendees.isEmpty() || request.getAttendees().isEmpty()
+        ? withOptionalAttendees
+        : getMeetingTimes(eventsCollection, request, /*includeOptionalAttendees=*/ false);
+  }
+
+  private Collection<TimeRange> getMeetingTimes(
+      Collection<Event> eventsCollection,
+      MeetingRequest request,
+      boolean includeOptionalAttendees) {
```

Comment:
Very Optional Nit: 

I'm really not a fan of boolean configuration params in functions. After maybe two, it starts to become very hard to mentally model the behavior due to the combinatorics of the booleans. I find it best to avoid them at all if possible.

In this case, you could do something like this:

```java

 private Collection<TimeRange> getAllAttendeesMeetingTimes(
      Collection<Event> events,
      MeetingRequest request) {
  HashSet<String> allAttendees = new HashSet<String>(request.getAttendees());
  attendees.addAll(request.getOptionalAttendees());
  return getMeetingTimes(request, getRelevantEvents(
       allAttendees, 
       new ArrayList<Event>(eventsCollection));
}

private Collection<TimeRange> getMandatoryAttendeesMeetingTimes(
      Collection<Event> events,
      MeetingRequest request) {
  return getMeetingTimes(request, getRelevantEvents(
       new HashSet<String>(request.getAttendees()), 
       new ArrayList<Event>(eventsCollection));
}

// N.B. I've reversed the params for "clean-ness" in call site, and to avoid
// people accidentally calling the wrong overload. Just a personal taste thing.
private Collection<TimeRange> getMeetingTimes(
    MeetingRequest request     
    Collection<Event> relevantEvents) {
  // The rest of the implementation
}
```

Reply:
Ack